rust/src/mqtt/parser.rs

```diff
@@ -0,0 +1,631 @@
+/* Copyright (C) 2020 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+// written by Sascha Steinbiss <sascha@steinbiss.name>
+
+use crate::mqtt::mqtt_message::*;
+use crate::mqtt::mqtt_property::*;
+use nom::combinator::rest;
+use nom::number::streaming::*;
+use nom::*;
+use std::str;
+
+#[derive(Debug)]
+pub struct FixedHeader {
+    pub message_type: u8,
+    pub dup_flag: bool,
+    pub qos_level: u8,
+    pub retain: bool,
+    pub remaining_length: u32,
+}
+
+// PARSING HELPERS
+
+#[inline]
+fn is_continuation_bit_set(b: u8) -> bool {
+    return (b & 128) != 0;
+}
+
+#[inline]
+fn convert_varint(continued: Vec<u8>, last: u8) -> u32 {
+    let mut multiplier = 1u32;
+    let mut value = 0u32;
+    for val in &continued {
+        value = value + ((val & 127) as u32 * multiplier);
+        multiplier = multiplier * 128u32;
+    }
+    value = value + ((last & 127) as u32 * multiplier);
+    return value;
+}
+
+#[inline]
+fn varint_length(val: usize) -> usize {
+  match val {
+      0 ..= 127 => 1,
+      128 ..= 16383 => 2,
+      16384 ..= 2097151 => 3,
+      2097152 ..= 268435455 => 4,
+      _ => 0,
+  }
+}
+
+// DATA TYPES
+
+named!(#[inline], pub parse_mqtt_string<String>,
+       do_parse!(
+           length: be_u16
+           >> content: take!(length)
+           >>  (
+                 str::from_utf8(&content).unwrap_or("NON-UTF8").to_string()
```

Comment:
What does `unwrap_or("NON-UTF8")` ?

Reply:
It tries to get the string from the `from_utf8` conversion function, and in case this fails, just returns the fixed string `NON-UTF8`. Since in MQTT valid strings are _always_ UTF8, this case would trigger only when parsing a non-compliant message. This code is just to avoid runtime issues with `unwrap`.

However, I can see that it might be better to just use `String::from_utf8_lossy` in this case, which would preserve the original string but just replace non-UTF8 characters with a replacement. See https://doc.rust-lang.org/std/string/struct.String.html#method.from_utf8_lossy for more info. I will use that in this case.

Reply:
Thanks for the explanation for myself :-)